### PR TITLE
Further VectorMemory cleanups.

### PR DESCRIPTION
**Pull request by bangerth**
While there, also convert SolverCG to a scheme where we use smart pointers
instead of things we have to allocate/deallocate by hand.

Follow-up to #4925.

### CODE REVIEW

**Comment by bangerth** [comment]
In reference to #4935.

File: include/deal.II/lac/vector_memory.templates.h [hunk 1]
```diff
@@ -42,18 +43,12 @@ inline
...
-    {
-      delete i->second;
-    }
+  if (data == nullptr)
+    return;
+
+  // delete the 'data' object. this also releases all vectors
+  // that are pointed to by the std::unique_ptrs
+  data->clear();
   delete data;
```
Comment (drwells) [inline include/deal.II/lac/vector_memory.templates.h]: Is there any reason why `data` is a pointer to a `std::vector` instead of just a `std::vector`?

Comment (bangerth) [inline include/deal.II/lac/vector_memory.templates.h]: I wondered about that as well. I don't recall, so I left that part alone. I have a suspicion that it had to do with explicitly managing when things are destroyed. For example, the destructor of that class does not delete the pointer. This creates a memory leak (but then the owner is a `static` object anyway, whose destructor is only run at the end of the program) that I can only imagine is on purpose. But I can not say that I would recall (or even remember who wrote this).

Comment (bangerth) [inline include/deal.II/lac/vector_memory.templates.h]: Actually, I take that back. This code is in the destructor, so the memory is released. 

But I still don't know the answer to your question. I decided to do one thing at a time :-)

Comment (drwells) [inline include/deal.II/lac/vector_memory.templates.h]: We can just leave it for later. This PR is still a nice clean up either way.

**Pull request merged**